core/src/main/java/be/orbinson/aem/dictionarytranslator/servlets/datasource/LanguageDatasource.java

```diff
@@ -19,67 +24,186 @@
-        ValueMap valueMap = new ValueMapDecorator(Map.of("value", language, "text", value + " (" + language + ")"));
-        resourceList.add(new ValueMapResource(resourceResolver, "", "", valueMap));
+    private static Map<String, String> toLanguageMap(List<ValueTextResource> resources) {
+        return resources.stream()
+                // the upstream data source does not filter access control child resource
+                .filter(r -> !AccessControlConstants.REP_POLICY.equals(r.getValue()))
+                .collect(Collectors.toMap(
+                ValueTextResource::getValue,
+                r -> r.getText() + " (" + r.getValue() + ")",
+                (oldValue, newValue) -> {
+                            LOG.warn("Duplicate language/country code: {}", oldValue);
+                            return oldValue;
+                }));
     }
 
     @Override
-    protected void doGet(SlingHttpServletRequest request, @NotNull SlingHttpServletResponse response) {
+    protected void doGet(@NotNull SlingHttpServletRequest request, @NotNull SlingHttpServletResponse response) throws ServletException, IOException {
+        // populate language map and filter
         String dictionaryPath = request.getRequestPathInfo().getSuffix();
+        if (dictionaryPath == null) {
+            throw new IllegalArgumentException("This data source must always be called with a dictionary path as request suffix");
+        }
+        Map<String, String> languageMap = getAllAvailableLanguages(request, response);
+        Set<String> dictionaryLanguages = getDictionaryLanguages(request.getResourceResolver(), dictionaryPath);
+        Predicate<String> languageFilter;
+        // evaluate data source configuration
+        Config dsCfg = new Config(request.getResource().getChild("datasource"));
+        if (dsCfg.get("hideNonDictionaryLanguages", false)) {
+            languageFilter = dictionaryLanguages::contains;
+            // add missing languages to dictionary
+            dictionaryLanguages.forEach(l -> languageMap.putIfAbsent(l, l));
+        } else {
+            languageFilter = l -> !dictionaryLanguages.contains(l);
+        }
+        // convert to list of resources
+        boolean emitTextFieldResources = dsCfg.get("emitTextFieldResources", false);
+        List<OrderedValueMapResource> resourceList = languageMap.entrySet().stream()
+                .filter(e -> languageFilter.test(e.getKey()))
+                .map(e -> {
+                    if (emitTextFieldResources) {
+                        return TextFieldResource.create(request.getLocale(), request.getResourceResolver(), e.getKey(), e.getValue());
+                    } else {
+                        return ValueTextResource.create(request.getLocale(), request.getResourceResolver(), e.getKey(), e.getValue());
+                    }
+                })
+                .collect(Collectors.toList());
+        // sort by display names
+        Collections.sort(resourceList);
```

Comment:
which sort order should be used:
a) by language/country name (label)
b) by language/country code

The former was and is used for "Create Language", "Delete Language", and "Create Key". The latter is used for "Edit" (https://github.com/orbinson/aem-dictionary-translator/pull/52). We should consolidate, but which one to pick?

Reply:
For me it's fine to use the same order as all the other orders in AEM, which one would that be?